app/gui/src/project-view/components/GraphEditor/widgets/WidgetSelection/SelectionSubmenu.vue

```diff
@@ -0,0 +1,139 @@
+  entries: Entry[]
+  selectedExpressions: Set<string>
+  topLevel?: boolean
+}>()
+
+const emit = defineEmits<{
+  clickedEntry: [Entry, boolean]
+}>()
+
+const dropdownElement = useTemplateRef('dropdownElement')
+const floatReference = computed(() => props.floatReference)
+const rootElement = computed(() => props.rootElement)
+
+const { floatingStyles } = submenuDropdownStyles(
+  floatReference,
+  dropdownElement,
+  props.topLevel,
+  rootElement,
+)
+
+/** Referring to the type of the component in the current file is hard, so we define a helper type. */
+interface SubmenuComponent {
+  isTargetOutside: (event: Event) => boolean
+}
+function isSubmenuComponent(component: unknown): component is SubmenuComponent {
+  return (
+    component != null &&
+    typeof component === 'object' &&
+    'isTargetOutside' in component &&
+    typeof component.isTargetOutside === 'function'
+  )
+}
+
+/** Check if the event target is outside the current submenu and any of its descendants. */
+function isTargetOutside(event: Event) {
+  const isOutsideCurrent = targetIsOutside(event, unrefElement(dropdownElement))
+  const isOutsideSubmenu =
+    isSubmenuComponent(submenuRef.value) ? submenuRef.value.isTargetOutside(event) : true
+  return isOutsideCurrent && isOutsideSubmenu
+}
+
+defineExpose({
+  isTargetOutside,
+})
+
+export interface Submenu {
+  entries: ComputedRef<Entry[]>
+  relativeTo: HTMLElement
+}
+
+const submenu = ref<Submenu | null>(null)
+const submenuEntries = computed(() => submenu.value?.entries ?? [])
+const submenuRef = useTemplateRef('submenuRef')
+
+function resetSubmenu() {
+  submenu.value = null
+}
+watch([() => props.show, () => props.entries], resetSubmenu)
+
+const nestedEntriesPresent = computed(() =>
+  props.entries.some((entry) => isEntry(entry) && entry.tag instanceof NestedChoiceTag),
+)
+
+function nestedChoiceTagToSubmenu(tag: NestedChoiceTag, target: HTMLElement): Submenu {
+  const isSelected = (tag: ExpressionTag | NestedChoiceTag) =>
+    tag instanceof ExpressionTag && props.selectedExpressions.has(tag.expression)
+  const choiceToEntry = (choice: ExpressionTag | NestedChoiceTag): Entry => ({
+    value: choice.label,
+    selected: isSelected(choice),
+    tag: choice,
+  })
+
+  return {
+    entries: computed(() => tag.choices.map(choiceToEntry) satisfies Entry[]),
+    relativeTo: target,
+  }
+}
+
+function onClick(entry: DropdownEntry, keepOpen: boolean, target: HTMLElement) {
+  if (!isEntry(entry)) return
+  const tag = entry.tag
+  if (tag instanceof NestedChoiceTag) {
+    submenu.value = nestedChoiceTagToSubmenu(tag, target)
+  } else {
+    emit('clickedEntry', entry, keepOpen)
+  }
+}
+
+function onScroll() {
+  submenu.value = null
+}
+</script>
+
+<template>
+  <Teleport v-if="props.rootElement" :to="props.rootElement">
+    <div ref="dropdownElement" :style="floatingStyles" class="SelectionSubmenu widgetOutOfLayout">
+      <SizeTransition height :duration="100">
+        <DropdownWidget
```

Comment:
I just realized we have `DropdownWidget` and `DropdownMenu` which, of course, don't use `DropdownWidget` at all.

I think the difference should be documented. Or these two could be merged? But the latter shall be rather a separate task.

Reply:
Yeah… We have a couple of really confusing names here:
```
components/FloatingSelectionMenu.vue
components/GraphEditor/widgets/WidgetSelection
components/GraphEditor/widgets/WidgetSelection/SelectionSubmenu.vue
components/GraphEditor/widgets/WidgetSelection.vue
components/SelectionDropdown.vue
components/SelectionDropdownText.vue
components/SelectionMenu.vue
components/DropdownMenu.vue
components/SelectionDropdown.vue
components/SelectionDropdownText.vue
components/widgets/DropdownWidget.vue
```

Reply:
I’m not sure how to address it properly at the moment, we might need to do some renamings/refactorings, but I think it is out of scope of this PR. I will create an issue instead.